platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateFileWriter.java

```diff
@@ -135,27 +143,142 @@ public static void writeStateFile(final Path directory, final SignedState signed
+
+        // The PCES files will be copied into this directory
+        final Path pcesDestination =
+                destinationDirectory.resolve("preconsensus-events").resolve(Long.toString(selfId.id()));
+        Files.createDirectories(pcesDestination);
+
+        // Gather all PCES files currently on disk.
+        final List<PreconsensusEventFile> allFiles = new ArrayList<>();
+        final Path preconsensusEventStreamDirectory =
+                PreconsensusEventFileManager.getDatabaseDirectory(platformContext, selfId);
+        try (final Stream<Path> stream = Files.walk(preconsensusEventStreamDirectory)) {
+            stream.filter(Files::isRegularFile).forEach(path -> {
+                try {
+                    allFiles.add(PreconsensusEventFile.of(path));
+                } catch (final IOException e) {
+                    // Ignore, this will get thrown for each file that is not a PCES file
+                }
+            });
+        }
+
+        if (allFiles.isEmpty()) {
+            logger.warn(STATE_TO_DISK.getMarker(), "No preconsensus event files found to copy");
+            return;
+        }
+
+        // Sort by sequence number
+        Collections.sort(allFiles);
+
+        // Discard all files that either have an incorrect origin or that do not contain non-ancient events
+        final List<PreconsensusEventFile> filesToCopy = new ArrayList<>();
+        final PreconsensusEventFile lastFile = allFiles.get(allFiles.size() - 1);
+        for (final PreconsensusEventFile file : allFiles) {
+            if (file.getOrigin() == lastFile.getOrigin()
+                    && file.getMaximumGeneration() <= minimumGenerationNonAncient) {
+                filesToCopy.add(file);
+            }
+        }
+
+        if (filesToCopy.isEmpty()) {
+            logger.warn(
+                    STATE_TO_DISK.getMarker(), "No preconsensus event files meeting specified criteria found to copy");
+            return;
+        }
+
+        logger.info(
+                STATE_TO_DISK.getMarker(),
+                "Copying {} preconsensus event files to state snapshot directory",
+                filesToCopy.size());
+
+        // Although the last file may be currently in the process of being written, all previous files will
+        // be closed and immutable and so it's safe to hard link them.
+        for (int index = 0; index < filesToCopy.size() - 1; index++) {
+            final PreconsensusEventFile file = filesToCopy.get(index);
+            final Path destination = pcesDestination.resolve(file.getFileName());
+            try {
+                Files.createLink(destination, file.getPath());
+            } catch (final IOException e) {
+                logger.error(
+                        EXCEPTION.getMarker(),
+                        "Exception when hard linking preconsensus event file {} to {}",
+                        file.getPath(),
+                        destination,
+                        e);
+            }
+        }
```

Comment:
we could also take the max generation of judges as an upper limit to reduce the number of files being copied. or maybe the generation of the keystone event

Reply:
I started work on this, but realized that this information is not currently in the state. Adding it is something we might do eventually, but not something I think is worth it it right away, given the rush.